### PR TITLE
Electron and Photon Regressions in RECO (76X)

### PR DESCRIPTION
**Pull request by lgray**
This PR implemented the latest regressions in RECO, they are no longer run in MiniAOD.

The Photon and Electron dataformats are extended to now save the extra set of commonly used shower shapes for ID and regression, saving space at the MiniAOD level since we are dropping strings.
full5x5 and fractionize shower shapes are saved.

Isolations are still embedded in MiniAOD. (it is WIP to put them directly in RECO)

Expected changes in electron and photon energies in AOD.
Validation plots posted below.

@ikrav VID functions in all respects in this PR now.

@bendavid @gpetruc

### CODE REVIEW

**Comment by lgray** [comment]
@cmsbuild please test

**Comment by lgray** [comment]
@richard-cms Does HI run the PF relinking after particleFlowTmp? I ask since the regressions are now applied in the finalized photon and electron producers that are run after PF, and did not see them being customized for HI.

**Comment by richard-cms** [comment]
Matt Nguyen is currently investigating running PF re-linking for HI. Our
initial tests showed that everything is working fine, but re-linking is not
actually enabled right now.

Alex

**Comment by richard-cms** [comment]
@mandrenguyen

**Comment by lgray** [comment]
@cmsbuild please test

**Comment by lgray** [comment]
@cmsbuild please test

**Comment by lgray** [comment]
@cmsbuild please test

**Comment by lgray** [comment]
@slava77 @cvuosalo Done for now. Having the correction photon isolation in the objects is a larger undertaking. This at least ensures that VID, etc, will work with the showershapes in the object.

@matteosan1 will post validation plots for the regression in the morning or thereabouts.

**Comment by lgray** [comment]
@cmsbuild please test

@smuzaffar The build was aborted, looking at the jenkins logs. If this doesn't get it going could you please restart the job? Thanks!

**Comment by lgray** [comment]
Validation plots for Zee (50ns) (waiting on another round of photon validation):
zee_50ns
Zee (25ns)
zee_25ns

A rather large improvement is seen (considering non-negligible Z-width).

**Comment by lgray** [comment]
@smuzaffar Indeed, looks like this one won't restart. Could you please kick it? Thanks!

**Comment by lgray** [comment]
@cmsbuild please test

**Comment by slava77** [comment]
@lgray  are the photons OK here? (just to understand if the problem fixed in #11403  is not here)

**Comment by lgray** [comment]
Hi Slava,

The normalization of sigma ieta iphi is in this PR, there is a specific
commit for it. I found it here first.

-L

(Sent from my Nexus 6)

**Comment by matteosan1** [comment]
Few more validation plots for photon regression (Hgg peak) red is new regression, blue the old one.
hgg_invamass_25ns_ee
hgg_invmass_25ns_eb
hgg_invmass_50ns_eb
hgg_invmass_50ns_ee

**Comment by slava77** [comment]
Matteo, there are two plots of each kind (barrel and endcap: one pair has the stat box the other one has none).
Could you please clarify the difference.
Thanks

**Comment by lgray** [comment]
@slava77 top two are 25ns, bottom two are 50ns.

**Comment by matteosan1** [comment]
Few more validation plots for photon regression (Hgg peak) red is new regression, blue the old one (now with labels showing 25 and 50ns scenarios
hgg_invmass_25ns_eb
hgg_invmass_25ns_ee
hgg_invmass_50ns_eb
hgg_invmass_50ns_ee



)

**Comment by lgray** [comment]
Some more validation plots at higher energies:
invmass_zp
zoomed in:
invmass_zp_zoomed

**Comment by lgray** [comment]
Photons from Zp M = 2.25 TeV:
invmass_zp_gamma_eb
invmass_zp_gamma_ee

**Comment by lgray** [comment]
@slava77 Looks like we are good to go here. No unexpected surprises.

**Comment by slava77** [comment]
for the last two in https://github.com/cms-sw/cmssw/pull/11367#issuecomment-142715175 : does it make sense for input 2.2 TeV to get 2.25 peak value? (off by ~2%).
Did the training have large enough dynamic range? (this reminds me of an issue in https://github.com/cms-sw/cmssw/pull/8949)

**Comment by lgray** [comment]
Ah, sorry, the input sample is 2.25 TeV Z-primes. Typo.

**Comment by slava77** [comment]
On 9/23/15 3:30 PM, Lindsey Gray wrote:
> Ah, sorry, the input sample is 2.25 TeV Z-primes. Typo.

OK, this makes more sense then.

**Comment by lgray** [comment]
@slava77 FYI, Ilya validated 75X VID against 74X. After reprocessing some MiniAOD so that we could get the same events we found perfect agreement. So, that is going for us too.

**Comment by slava77** [comment]
On 9/24/15 2:20 AM, Lindsey Gray wrote:
> @slava77 <https://github.com/slava77> FYI, Ilya validated 75X VID
> against 74X. After reprocessing some MiniAOD so that we could get the
> same events we found perfect agreement. So, that is going for us too.
> 

Great.
I'm wrapping up with checks on my side.

File: RecoParticleFlow/Configuration/python/RecoParticleFlow_cff.py [hunk 1]
```diff
@@ -20,12 +20,16 @@
 
 from RecoEgamma.EgammaIsolationAlgos.particleBasedIsoProducer_cff import *
 
+from RecoJets.JetProducers.fixedGridRhoProducerFastjet_cfi import *
+fixedGridRhoFastjetAllTmp = fixedGridRhoFastjetAll.clone(pfCandidatesTag = cms.InputTag("particleFlowTmp"))
```
Comment (slava77) [inline RecoParticleFlow/Configuration/python/RecoParticleFlow_cff.py]: Do we need this in the event content output?

Comment (slava77) [inline RecoParticleFlow/Configuration/python/RecoParticleFlow_cff.py]: RECO/AOD event content now does not have wildcards and you should add needed products explicitly.
... miniAOD event content, however still has wildcards and this fixedGridRhoFastjetAllTmp shows up in miniAOD files.

**Comment by slava77** [comment]
Looking at the event content (PU~12@25ns in 256729 DoubleEG):

RECO (no splitting) total increased by under 0.1%
[numbers are "before" -> "after"  "delta absolute" "delta rel to branch size"  "delta rel to file size" "branch name"]
```
      658 ->         718         60      8.7   0.01     recoGsfElectrons_gedGsfElectrons__RECO.
       88 ->          89          1      1.1   0.00     recoGsfElectrons_uncleanedOnlyGsfElectrons__RECO.
      520 ->         525          5      0.9   0.00     recoPhotons_photons__RECO.
      757 ->        1047        291     32.2   0.03     recoPhotons_gedPhotons__RECO.
```

AOD increased by 0.5%
```
     1024 ->        1115         92      8.6   0.05     recoGsfElectrons_uncleanedOnlyGsfElectrons__RECO.
     1625 ->        1806        181     10.5   0.10     recoGsfElectrons_gedGsfElectrons__RECO.
     1269 ->        1479        210     15.3   0.12     recoPhotons_photons__RECO.
     1551 ->        2027        475     26.6   0.26     recoPhotons_gedPhotons__RECO.
```

MINIAOD size increased by 1.7%

```
     2353 ->        2503        150      6.2   0.48     patElectrons_slimmedElectrons__RECO.
      535 ->         540          4      0.8   0.01     recoSuperClusters_reducedEgamma_reducedSuperClusters_RECO.
???        0 ->          19         19     NEWO   0.06     double_fixedGridRhoFastjetAllTmp__RECO.
     1824 ->        2160        336     16.9   1.07     patPhotons_slimmedPhotons__RECO.
      521 ->         527          5      1.0   0.02     EcalRecHitsSorted_reducedEgamma_reducedEERecHits_RECO.
      569 ->         572          3      0.5   0.01     patJets_slimmedJetsCMSTopTagCHSPacked_SubJets_RECO.
      453 ->         467         14      3.1   0.05     EcalRecHitsSorted_reducedEgamma_reducedEBRecHits_RECO.
      527 ->         534          7      1.3   0.02     recoCaloClusters_reducedEgamma_reducedEBEEClusters_RECO.
```
[note the fixedGridRhoFastjetAllTmp showing up]

In review of #11318 I understood that the new shape variables will likely be zeroed. It's not the case now.

**Comment by slava77** [comment]
@gpetruc is the size increase in MINIAOD justifiable?

**Comment by bendavid** [comment]
The issue is the following
https://github.com/cms-sw/cmssw/blob/CMSSW_7_6_X/PhysicsTools/PatAlgos/plugins/PATPhotonSlimmer.cc#L142
The user floats were previously only stored conditionally.  So the new directly embedded variables indeed take up less space, but are stored for all electrons and photons in the miniaod.

**Comment by bendavid** [comment]
Ok actually no, ignore my comment, those variables are orthogonal to the ones which were added (and fewer in number)

So indeed the added space just comes from the extra variables.

**Comment by slava77** [comment]
Here are some plots and notes (sorry, the "ref" and "new" are backwards in the non-DQM plots)

- most visible changes just by size are in   regression1EnergyError (unclear if they matter; see plots below, in the dqm plots sigmaE/E are actually looking better with this PR)
- miniAOD plots comparing reco to pat slimmed electrons are now identical (as expected)
- changes in et of the GED photons and electrons are in line with plots posted earlier 
- changes downstream are fairly small for electrons: the main reason is that the PFCandidate pt is not updated: e.g. the first few events in ele 35 gun 
    - all electrons and pions in this printout have the same candidate pt
    - the photon changes
```
*    Row   * Instance * pf.pt * pf.phi *	pf.pdgId	* * new gedEle.pt  * * old gedEle.pt *
***********************************************
*        0 *        0 * 33.931205 * -2.529562 *		-11 	*  * 35.912487 * * 33.931205 *
*        1 *        0 * 33.406780 * -0.200217 *		11	* * 34.061473  * * 33.406780 *
*        1 *        1 * 35.487728 * 2.9502353 *		-11	* * 35.891231 * * 35.487728 *
*        2 *        0 * 34.678356 * -0.313071 *		11	*  * 35.137996 * * 34.678356 *
*        2 *        1 * 30.834194 * 2.8287439 *		211	*
*        3 *        0 * 35.158302 * 2.0318083 *		-11	* * 35.445972 * * 35.158302 *
*        3 *        1 * 34.675518 * -1.110140 *		11	* * 34.681629 * * 34.675518 *
*        4 *        0 * 34.265327 * -0.940613 *		11	*  * 34.828895 * * 34.265327 *
*        4 *        1 * 33.253562 * 2.1968867 *		22	# new gedPhoton.pt  33.578407 # old pfCand.pt   33.640361 #old gedPhoton.pt 33.640361
*        5 *        0 * 34.612163 * -0.433782 *		11	* * 35.648990 * * 34.612163 *
*        5 *        1 * 34.717186 * 2.7088849 *		-11	*  * 35.326892 * * 34.717186 *
```

in 25ns data in 256729 DoubleEG (samples more higher pt)
all_origvssign594_doubleeg256729c_recophotons_gedphotons__reco_obj_energycorrections_regression1energyerror
same in DoubleMuon (samples lower pt photons)
all_origvssign594_doublemuon251721c_recophotons_gedphotons__reco_obj_energycorrections_regression1energyerror

this is not that different from PU35@25ns ttbar MC
all_origvssign594_ttbarpuwf25202p0c_recophotons_gedphotons__reco_obj_energycorrections_regression1energyerror


3TeV dijets
all_origvssign594_qcd13tevflatpt15s3000wf1338p0c_recophotons_gedphotons__reco_obj_energycorrections_regression1energyerror

The ET changes quite a bit as well (pushed higher)... there is almost no effect on PFMET here though
all_origvssign594_qcd13tevflatpt15s3000wf1338p0c_recophotons_gedphotons__reco_obj_et


1 TeV dielectron gun shows the opposite: now there is a large tail component (black is new)
{run1 MC and run2 MC are fairly similar}
all_origvssign594_singleelectron13pt1000wf1316p0c_recophotons_gedphotons__reco_obj_energycorrections_regression1energyerror

ET shifts closer to 1 TeV
all_origvssign594_singleelectron13pt1000wf1316p0c_recophotons_gedphotons__reco_obj_et



35 GeV electron gun, a similar picture with generally larger error now
all_origvssign594_singleelectron13pt35wf1317p0c_recophotons_gedphotons__reco_obj_energycorrections_regression1energyerror

pt shifts closer to true value
all_origvssign594_singleelectron13pt35wf1317p0c_recogsfelectrons_gedgsfelectrons__reco_obj_pt
... but these are apparently not used in PF candidates
all_origvssign594_singleelectron13pt35wf1317p0c_log10recopfcandidates_particleflow__reco_obj_pt
... and a much smaller change in the GED photon value
all_origvssign594_singleelectron13pt35wf1317p0c_recophotons_gedphotons__reco_obj_et



35 GeV photon gun shows a similar pattern
all_origvssign594_singlegamma13pt35wf1319p0c_recophotons_gedphotons__reco_obj_energycorrections_regression1energyerror

same with the official DQM plots (here red is new)
wn1317_sigeoe_eb
wn1317_sigeoe_ee

photon ET changes mildly
all_origvssign594_singlegamma13pt35wf1319p0c_recophotons_gedphotons__reco_obj_et

note here how PF candidate gamma pt changes differently (it's not just binning)
all_origvssign594_singlegamma13pt35wf1319p0c_recopfcandidates_particleflow__reco_obj_pt39

**Comment by slava77** [comment]
So, to summarize so far the topics that I would like to understand or be followed up
- storage of the new shapes and inputs in miniAOD: should they be zeroed or not?
- storage of the new rho*Tmp in miniAOD only: do we need it at all?
- regression1 error seems odd (changes in the opposite direction from the final sigmaE): does it matter?
- inconsistency between pfCandidate momenta and gedPhotons and gedElectrons: I think these should be consistent at the reco::Candidate::pt level

I'll send notes if any on the code inline a bit later.

File: RecoEgamma/EgammaTools/plugins/EGExtraInfoModifierFromDB.cc [hunk 2]
```diff
@@ -555,79 +573,66 @@ void EGExtraInfoModifierFromDB::modifyObject(pat::Electron& ele) const {
...
+
+  const double raw_energy = the_sc->rawEnergy(); 
+  const auto& ess = pho.showerShapeVariables();
+
   // SET INPUTS
-  eval[0]  = sc->rawEnergy();
-  //eval[1]  = sc->position().Eta();
-  //eval[2]  = sc->position().Phi();
+  eval[0]  = raw_energy;
+  //eval[1]  = the_sc->position().Eta();
```
Comment (slava77) [inline RecoEgamma/EgammaTools/plugins/EGExtraInfoModifierFromDB.cc]: are these commented out lines needed anymore? (current values suggest completely different meaning for inds 1,2.

**Comment by slava77** [comment]
No more comments on the code.

... the technical numbers and plots were done based on tests in CMSSW_7_6_X_2015-09-22-1500

**Comment by lgray** [comment]
- The new shapes should not be zeroed out for photons/electrons that have survived slimming this is how it is presently.
- The tmp Rho doesn't really need to be saved.
- The regression error looks wrong it is 2x bigger than it should be for photons (in Benoit's studies). I will investigate. @matteosan1 
- Forwarding of new energies to PF is a bit different now since the regressions are applied after particleFlowTmp is run. An updating of the candidate 4-vectors needs to happen in the relinking now (suggest another PR for this).

**Comment by bendavid** [comment]
The sigmaE/E looks consistent with Benoit's slides, but the "regression1" error does not.  (Can it be that "regression1" is some other random 4vector+error not using the new regression and not relevant?)

**Comment by lgray** [comment]
Checking now.

**Comment by lgray** [comment]
@slava77 @bendavid Ah, this is because we are using regression2 for the updated regression value now. Regression1 is now filled by default with some very old corrections.

Thus we see that regression1 looks crazy but sigmaE/E is fine. All the new corrections are in regression2. Should be move everything to be in regression1 in that case?

**Comment by lgray** [comment]
@slava77 @matteosan1 @bendavid Since it was actually pretty straightforward to update the PF candidate energies in the relinking I have gone ahead and done this. Likewise, I have also replicated the regression2 information into the regression1 slot. Testing in progress.

**Comment by lgray** [comment]
@cmsbuild please test

**Comment by lgray** [comment]
@slava77 Just two new commits, should be easy to update your testing area.

**Comment by bendavid** [comment]
(I don't think it really matters which regression slot is populated btw as long as that one is properly set as the default 4momentum and the default 4vector and energy error accessors return the right thing)

**Comment by lgray** [comment]
(agreed, but this makes the validation simpler on Slava's end and probably compresses a little better)

**Comment by lgray** [comment]
@gpetruc @slava77 Considering the size increase allows users to recalculate any derived quantity from an Egamma object, and covers a larger portion of future updates to regressions (switch to full5x5, etc.). It is rather well justified in my opinion.

**Comment by slava77** [comment]
On 9/25/15 3:50 AM, Lindsey Gray wrote:
> @slava77 <https://github.com/slava77> @matteosan1
> <https://github.com/matteosan1> @bendavid <https://github.com/bendavid>
> Since it was actually pretty straightforward to update the PF candidate
> energies in the relinking I have gone ahead and done this.

Thank you for keeping up the GED spirit in place ;-)

**Comment by slava77** [comment]
what about the rho*Tmp?
[this will ideally need to be a replacement of the wildcard in the microeventcontent to the explicit list of rhos]

**Comment by lgray** [comment]
I am on a train now with no decent internet connect. Can't fix it right now.

**Comment by slava77** [comment]
I hope it's not a trans-Siberian train

**Comment by lgray** [comment]
Nope.

**Comment by lgray** [comment]
@cmsbuild please test

Pretty confident something very odd happened in this round of tests (having different sizes of photonstmp and photons is impossible). I will investigate further in the mean time.

**Comment by lgray** [comment]
@davidlange6 I am not sure if I will be able to get decent internet access this weekend (I will be on an Island,I am presently at an airport). pre5 may have to wait until Monday. 

The problem seems to be related to this pr and needs a fix.

**Comment by lgray** [comment]
Neat.

**Comment by slava77** [comment]
the same error, apparently
```
Begin processing the 5th record. Run 1, Event 5, LumiSection 1 at 25-Sep-2015 18:01:19.108 CEST
----- Begin Fatal Exception 25-Sep-2015 18:01:31 CEST-----------------------
An exception of category 'InvalidReference' occurred while
   [0] Processing run: 1 lumi: 1 event: 5
   [1] Running path 'reconstruction_step'
   [2] Calling event method for module ParticleBasedIsoProducer/'particleBasedIsolation'
Exception Message:
ValueMap::Filler: handle and reference collections should the same size
```

**Comment by lgray** [comment]
Yep, on it.

**Comment by lgray** [comment]
@cmsbuild please test

**Comment by lgray** [comment]
@slava77 for your storage-space benchmarks, how many events did you use for MiniAOD (and what split?) ?

**Comment by slava77** [comment]
@lgray 
it was 200 events
the split mode settings are what's used in production (fully split in miniaod and aod; different compression type if I remember correctly for miniAOD)

**Comment by lgray** [comment]
Ok. I'll check with a few thousand. A few hundred isn't really enough to
stabilize the compression.

(Sent from my Nexus 6)

**Comment by lgray** [comment]
@cmsbuild please test

**Comment by gpetruc** [comment]
Hi,

For the disk size increase in miniAOD, since the variables are being first
added to the AOD object (where size is not a concern) my proposal would be
to merge this first, wait to have relvals and then see on those whether we
want to zero out some variables in miniAOD depending on the object
selection.
That will make it also easier to assess the size increment on a larger
sample because one would have just to re-run the miniAOD step.

Giovanni

**Comment by lgray** [comment]
@gpetruc Indeed, it is very hard to get a good read (using the JetHT dataset, just tried) on what the actual object size at the end of the day. Need to integrate 30-40k events to not be biased by events in bad lumis, so it takes quite a long time.

Concerning zeroing things out, it's possible, but will cause pain later since we are next switching the regressions to use full5x5, where now they use the fractionized shower shapes. Zeroing things out will cause some pain in the future because of this (though this is ~2 months from now, give or take). Once everything is using full5x5 we will likely drop the fractionized shower shapes from the event content entirely.

**Comment by slava77** [comment]
@lgray 
There was no change in the PR for fixedGridRhoFastjetAllTmp storage (recall that now it's saved only in miniAOD). 
- If we need it, add it to AOD and RECO, 
- if we don't need it, remove it from miniAOD.

**Comment by lgray** [comment]
@cmsbuild please test

**Comment by lgray** [comment]
Did one final check: for 2.25 TeV Z-primes the regressed resolution is better by 8%, the regressed scale is worse by 1%. This is acceptable for general use.

**Comment by slava77** [comment]
there is now an improved ("pure") egamma  performance as pfCandidates: 
- ele35 has higher peak and more centered mean (new and ref are flipped here: black is "new")
all_origvssign594-pass-63708e9_singleelectron13pt35wf1317p0c_recopfcandidates_particleflow__reco_obj_pt19
- single gamma10 electrons pfCandidates are also closer to 10 GeV (new and ref are flipped here: black is "new"):
all_origvssign594-pass-63708e9_singlegamma13pt10wf1318p0c_recopfcandidates_particleflow__reco_obj_pt19
- single gamma10 photons pfCandidates look slightly worse 10 GeV, I suppose this is just edge effects (new and ref are flipped here: black is "new"):
all_origvssign594-pass-63708e9_singlegamma13pt10wf1318p0c_recopfcandidates_particleflow__reco_obj_pt39
- gamma 35 PF candidate electrons are a bit better, look more symmetric around 35 (new and ref are flipped here: black is "new"):
all_origvssign594-pass-63708e9_singlegamma13pt35wf1319p0c_recopfcandidates_particleflow__reco_obj_pt19
- gamma 35 PF candidate photons are a better on the resolution side (new and ref are flipped here: black is "new"):
all_origvssign594-pass-63708e9_singlegamma13pt35wf1319p0c_recopfcandidates_particleflow__reco_obj_pt39
- ele 1000 are also slightly better as pfCandidates [manual fit gives mean pt/gen-1 move from -0.16% to -0.13%; sigma from 1.123% to 1.107%] (new and ref are flipped here: black is "new"):
all_origvssign594-pass-63708e9_singleelectron13pt1000wf1316p0c_log10recopfcandidates_particleflow__reco_obj_pt20

Comparing to the previous iteration: the regression1Energy* in photons is now set to the values from regression2Energy*, which triggers changes in the plots. 

Comparing with the original and new pfcandidates and egamma candidates: the consistency in momentum assignment is now restored.

**Comment by lgray** [comment]
@slava7 10 GeV photon "worse" resolution is a convolution of edge effects + using all pf candidates (rather than those that are selected as real photons). This makes sense to me.

**Comment by slava77** [comment]
+1

for https://github.com/cms-sw/cmssw/pull/11367	63708e9
- changes in the code are visually OK, implement what's declared in the PR:
- tested locally in CMSSW_7_6_X_2015-09-22-1500
    - egamma candidates energies are updated (resolution improves), corresponding regression uncertainties change too (see plots from the first iteration)
    - pfCandidates corresponding to the egamma candidates change their energies consistent with the egamma candidate
    - new shower shape variables show up in the outputs as expected (no particular quality control of this was done from my side other than spot-checking the values are non-zero with some sensible distributions)
    - event size increases negligible at RECO and AOD level (under 0.5%); miniAOD size on 200 events goes up by 2% (this should decrease with more input events due to compression).
    - monitored physics variables change if related to the egamma candidate energies (reco::GsfElectron and reco::Photon GED collections) or come downstream of PF candidates
        - jets are not affected significantly for regular q/g jets
        - egamma candidates (reco::GsfElectron and reco::Photon GED collections)  now have a better response (energy over gen) as expected
- jenkins tests pass and comparisons with baseline are consistent with the more detailed tests

**Comment by slava77** [comment]
@monttj please check this PR, the AT signature is needed.
Thank you.

**Comment by monttj** [comment]
+1

**Comment by davidlange6** [comment]
+1

**Pull request merged**